### PR TITLE
v35: Add an action to setup dev tooling

### PR DESCRIPTION
**Pull request by olix0r**
* The _action_ image is removed.
* The _tools_ image now holds only the binaries we need for CI.
* The _runtime_ image is renamed to _devcontainer_ for clarity.
* `actions/setup-tools` extracts binaries for the tools image
* `actions/setup-go` Sets up the Go toolchain to match the devcontainer
* `actions/setup-rust` Sets up the Rust toolchain to match the devcontainer

Used in https://github.com/olix0r/kubert/pull/96

Signed-off-by: Oliver Gould <ver@buoyant.io>

### CODE REVIEW

**Comment by olix0r** [comment]
- [ ] TODO: we need our `just-dev` tooling to be able to validate that workflow dev tags match the image tag.

File: .github/workflows/actions.yml [hunk 1]
```diff
@@ -16,22 +17,49 @@ jobs:
...
+      - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+      - uses: ./actions/setup-tools
+      - run: which just
+      - run: just --version
+      - run: k3d version
+      - run: kubectl version --output=yaml --client
+      - run: just-cargo --list
+      - run: just-dev --list
+      - run: just-k3d --list
+      - run: just-sh --list
```
Comment (alpeb) [inline .github/workflows/actions.yml:49]: There's some repetition here on what `./actions/setup-tools` calls

**Comment by alpeb** [comment]
> * `actions/setup-go` Sets up the Go toolchain to match the devcontainer
> * `actions/setup-rust` Sets up the Rust toolchain to match the devcontainer

In the Dockerfile's `go` and `rust` stages we're also building binaries for go and rust. Should those be used, or the `setup` actions?

**Comment by alpeb** [comment]
It occurs to me we could avoid having to keep in sync the go/rust versions between the setup actions and the devcontainer by passing around the version from the setup actions. For example for go, something like this:
- The `setup-go` action would declare the go version, becoming the single source of truth, and would return that version as an output.
- A new `build-go` recipe accepts the go version as an arg. The `dev-build` job could first call `setup-go` to retrieve the version and pass it to that recipe.
- That recipe passes the version as an arg to the build command
- The `go` stage consumes it through an ARG:
```
ARG  GO_VERSION
FROM docker.io/golang:${GO_VERSION} as go
```

File: README.md [hunk 2]
```diff
@@ -92,3 +74,132 @@ some common build and testing tasks. Most of these scripts are implemented with
...
+      - run: just-cargo build
+```
+
+Build failures are automatically reported as annotations in the GitHub UI.
+
+#### `actions/setup-rust`
+
+The above example is fine when we're just building software, but when we want to
+start networked services in Docker (namely, k3d clusters), the containerized
+approach no longer works.
```
Comment (alpeb) [inline README.md:169]: Paragraph in wrong section?

**Comment by olix0r** [comment]
@alpeb

> In the Dockerfile's go and rust stages we're also building binaries for go and rust. Should those be used, or the setup actions?

We're not _building_ binaries. We use the go/rust base images. My initial attempt at this extracted those images into the runner, but that's a bit too slow for my liking (>1m). The actions/setup-go action is maintained by GitHub so we have pretty high confidence about its provenance and they pre-load these toolchains on runners so it's _much_ faster to use. Our Rust setup action is pretty minimal.

> setup actions are for CI, the devcontainer for development environments.

Sort of.

setup actions are used for CI specifically when you want the workflow to run on the runner directly (or in a specific docker image). The rust and go docker images can be used directly in a workflow (instead of the setup actions) when you don't care about access to the host network. Specifically, a containerized workflow/action can't access kubernetes when run in k3d, since the container's network is unable to reach the k8s API.

The Rust and Go images should _also_ be used for base images in our Dockerfiles, so all tool versions are pinned across the board.

> It occurs to me we could avoid having to keep in sync the go/rust versions between the setup actions and the devcontainer by passing around the version from the setup actions.

But then this diverges with what's used in development. I think it's simplest if we pin the versions across the board (i.e. v35 uses Rust 1.64 and Go 1.18.7, etc) versus letting the vary. The Go/Rust version overrides are mostly there for when we want to test compatibility with an older version (i.e. for libraries).

File: README.md (hunk 2 above)
Comment (olix0r) [reply README.md:169]: This is describing how the setup action is needed when the rust container isn't used directly.

**Comment by alpeb** [comment]
Thanks for the clarifications, makes better sense now :+1:

**Pull request merged**